File: sdk/go/trust.go
package foodblock

import (
	"math"
	"strings"
	"time"
)

// DefaultWeights are the default trust computation weights (Section 6.3).
var DefaultWeights = map[string]float64{
	"authority_certs": 3.0,
	"peer_reviews":    1.0,
	"chain_depth":     2.0,
	"verified_orders": 1.5,
	"account_age":     0.5,
}

// PeerReviewResult holds the peer review sub-score.
type PeerReviewResult struct {
	Count         int     `json:"count"`
	AvgScore      float64 `json:"avg_score"`
	WeightedScore float64 `json:"weighted_score"`
}

// TrustInputs holds the five raw trust inputs.
type TrustInputs struct {
	AuthorityCerts int              `json:"authority_certs"`
	PeerReviews    PeerReviewResult `json:"peer_reviews"`
	ChainDepth     int              `json:"chain_depth"`
	VerifiedOrders int              `json:"verified_orders"`
	AccountAge     float64          `json:"account_age"`
}

// TrustResult is the output of ComputeTrust.
type TrustResult struct {
	Score        float64     `json:"score"`
	Inputs       TrustInputs `json:"inputs"`
	MeetsMinimum bool        `json:"meets_minimum"`
}

// TrustBlock extends Block with optional metadata used by trust computation.
type TrustBlock struct {
	Block
	AuthorHash string `json:"author_hash,omitempty"`
	CreatedAt  string `json:"created_at,omitempty"`
}

// ComputeTrust computes a trust score for an actor from five inputs
// derived from the FoodBlock graph. Supports custom trust policies.
func ComputeTrust(actorHash string, blocks []TrustBlock, policy map[string]interface{}) TrustResult {
	if actorHash == "" {
		panic("FoodBlock: actorHash is required")
	}

	weights := mergeWeights(policy)
	now := time.Now()

	var requiredAuthorities []string
	if ra, ok := policy["required_authorities"]; ok {
		if arr, ok := ra.([]string); ok {
			requiredAuthorities = arr
		}
	}

	inputs := TrustInputs{
		AuthorityCerts: countAuthorityCerts(actorHash, blocks, requiredAuthorities),
		PeerReviews:    computePeerReviews(actorHash, blocks),
		ChainDepth:     computeChainDepth(actorHash, blocks),
		VerifiedOrders: countVerifiedOrders(actorHash, blocks),
		AccountAge:     computeAccountAge(actorHash, blocks, now),
	}

	score :=
		float64(inputs.AuthorityCerts)*weights["authority_certs"] +
			inputs.PeerReviews.WeightedScore*weights["peer_reviews"] +
			float64(inputs.ChainDepth)*weights["chain_depth"] +
			float64(inputs.VerifiedOrders)*weights["verified_orders"] +
			inputs.AccountAge*weights["account_age"]

	minScore := 0.0
	if ms, ok := policy["min_score"]; ok {
		switch v := ms.(type) {
		case float64:
			minScore = v
		case int:
			minScore = float64(v)
		}
	}

	return TrustResult{
		Score:        score,
		Inputs:       inputs,
		MeetsMinimum: score >= minScore,
	}
}

// ConnectionDensity measures connection density between two actors (Section 6.3 sybil resistance).
// Returns 0..1 where 0 = no shared refs, 1 = fully connected.
func ConnectionDensity(actorA, actorB string, blocks []TrustBlock) float64 {
	if actorA == "" || actorB == "" {
		return 0
	}

	refsA := make(map[string]bool)
	refsB := make(map[string]bool)

	for _, b := range blocks {
		if b.Refs == nil {
			continue
		}
		vals := flattenRefValues(b.Refs)
		containsA := containsStr(vals, actorA)
		containsB := containsStr(vals, actorB)

		if containsA {
			for _, v := range vals {
				if v != actorA {
					refsA[v] = true
				}
			}
		}
		if containsB {
			for _, v := range vals {
				if v != actorB {
					refsB[v] = true
				}
			}
		}
	}

	if len(refsA) == 0 || len(refsB) == 0 {
		return 0
	}

	shared := 0
	for ref := range refsA {
		if refsB[ref] {
			shared++
		}
	}

	union := make(map[string]bool)
	for ref := range refsA {
		union[ref] = true
	}
	for ref := range refsB {
		union[ref] = true
	}

	if len(union) == 0 {
		return 0
	}
	return float64(shared) / float64(len(union))
}

// CreateTrustPolicy creates a trust policy block.
func CreateTrustPolicy(name string, weights map[string]interface{}, opts map[string]interface{}) Block {
	state := map[string]interface{}{
		"name":    name,
		"weights": weights,
	}
	if opts != nil {
		if ra, ok := opts["required_authorities"]; ok {
			state["required_authorities"] = ra
		}
		if ms, ok := opts["min_score"]; ok {
			state["min_score"] = ms
		}
	}

	refs := map[string]interface{}{}
	if opts != nil {
		if author, ok := opts["author"]; ok {
			if s, ok := author.(string); ok {
				refs["author"] = s
			}
		}
	}

	return Create("observe.trust_policy", state, refs)
}

func mergeWeights(policy map[string]interface{}) map[string]float64 {
	result := make(map[string]float64)
	for k, v := range DefaultWeights {
		result[k] = v
	}
	if policy == nil {
		return result
	}
	if pw, ok := policy["weights"]; ok {
		switch w := pw.(type) {
		case map[string]interface{}:
			for k, v := range w {
				switch n := v.(type) {
				case float64:
					result[k] = n
				case int:
					result[k] = float64(n)
				}
			}
		case map[string]float64:
			for k, v := range w {
				result[k] = v
			}
		}
	}
	return result
}

func countAuthorityCerts(actorHash string, blocks []TrustBlock, requiredAuthorities []string) int {
	count := 0
	for _, b := range blocks {
		if b.Type != "observe.certification" {
			continue
		}
		if b.Refs == nil {
			continue
		}
		subject, _ := b.Refs["subject"].(string)
		if subject != actorHash {
			continue
		}
		if vu, ok := b.State["valid_until"].(string); ok {
			t, err := time.Parse(time.RFC3339, vu)
			if err != nil {
				t, err = time.Parse("2006-01-02", vu)
			}
			if err == nil && t.Before(time.Now()) {
				continue
			}
		}
		count++
	}
	return count
}

func computePeerReviews(actorHash string, blocks []TrustBlock) PeerReviewResult {
	var reviews []TrustBlock
	for _, b := range blocks {
		if b.Type != "observe.review" {
			continue
		}
		if b.Refs == nil {
			continue
		}
		subject, _ := b.Refs["subject"].(string)
		if subject != actorHash {
			continue
		}
		if _, ok := b.State["rating"]; !ok {
			continue
		}
		reviews = append(reviews, b)
	}

	if len(reviews) == 0 {
		return PeerReviewResult{}
	}

	totalWeighted := 0.0
	totalWeight := 0.0

	for _, review := range reviews {
		reviewerHash := ""
		if ah, ok := review.Refs["author"].(string); ok {
			reviewerHash = ah
		} else {
			reviewerHash = review.AuthorHash
		}
		density := ConnectionDensity(reviewerHash, actorHash, blocks)
		weight := 1 - density
		rating := toFloat64(review.State["rating"])
		totalWeighted += (rating / 5.0) * weight
		totalWeight += weight
	}

	sum := 0.0
	for _, r := range reviews {
		sum += toFloat64(r.State["rating"])
	}
	avgScore := sum / float64(len(reviews))

	weightedScore := 0.0
	if totalWeight > 0 {
		weightedScore = totalWeighted / totalWeight * float64(len(reviews))
	}

	return PeerReviewResult{
		Count:         len(reviews),
		AvgScore:      avgScore,
		WeightedScore: weightedScore,
	}
}

func computeChainDepth(actorHash string, blocks []TrustBlock) int {
	authors := make(map[string]bool)
	for _, b := range blocks {
		if b.Refs == nil {
			continue
		}
		refsActor := false
		for _, v := range b.Refs {
			switch val := v.(type) {
			case string:
				if val == actorHash {
					refsActor = true
				}
			case []interface{}:
				for _, item := range val {
					if s, ok := item.(string); ok && s == actorHash {
						refsActor = true
					}
				}
			}
		}
		if refsActor && b.AuthorHash != "" {
			authors[b.AuthorHash] = true
		}
	}
	return len(authors)
}

func countVerifiedOrders(actorHash string, blocks []TrustBlock) int {
	count := 0
	for _, b := range blocks {
		if !strings.HasPrefix(b.Type, "transfer.order") {
			continue
		}
		if b.Refs == nil {
			continue
		}
		buyer, _ := b.Refs["buyer"].(string)
		seller, _ := b.Refs["seller"].(string)
		if buyer != actorHash && seller != actorHash {
			continue
		}
		_, hasAdapterRef := b.State["adapter_ref"]
		_, hasPaymentRef := b.State["payment_ref"]
		if hasAdapterRef || hasPaymentRef {
			count++
		}
	}
	return count
}

func computeAccountAge(actorHash string, blocks []TrustBlock, now time.Time) float64 {
	for _, b := range blocks {
		if b.Hash == actorHash && b.CreatedAt != "" {
			t, err := time.Parse(time.RFC3339, b.CreatedAt)
			if err != nil {
				t, err = time.Parse("2006-01-02T15:04:05.000Z", b.CreatedAt)
			}
			if err != nil {
				continue
			}
			days := now.Sub(t).Hours() / 24
			return math.Min(days, 365)
		}
	}
	return 0
}

func flattenRefValues(refs map[string]interface{}) []string {
	var result []string
	for _, v := range refs {
		switch val := v.(type) {
		case string:
			result = append(result, val)
		case []interface{}:
			for _, item := range val {
				if s, ok := item.(string); ok {
					result = append(result, s)
				}
			}
		}
	}
	return result
}

func containsStr(slice []string, s string) bool {
	for _, v := range slice {
		if v == s {
			return true
		}
	}
	return false
}

func toFloat64(v interface{}) float64 {
	switch n := v.(type) {
	case float64:
		return n
	case int:
		return float64(n)
	case int64:
		return float64(n)
	}
	return 0
}

File: sdk/javascript/src/trust.js
/**
 * FoodBlock Trust Computation (Section 6.3)
 *
 * Computes a trust score for an actor from five inputs derived
 * from the FoodBlock graph. Supports custom trust policies.
 */

const DEFAULT_WEIGHTS = {
  authority_certs: 3.0,
  peer_reviews: 1.0,
  chain_depth: 2.0,
  verified_orders: 1.5,
  account_age: 0.5
}

/**
 * Compute trust score for an actor.
 *
 * @param {string} actorHash - Hash of the actor to score
 * @param {object[]} blocks - All known blocks (or relevant subset)
 * @param {object} [policy] - Trust policy override (weights, required_authorities, min_score)
 * @returns {object} { score, inputs, meets_minimum }
 */
function computeTrust(actorHash, blocks, policy = {}) {
  if (!actorHash || typeof actorHash !== 'string') {
    throw new Error('FoodBlock: actorHash is required')
  }
  if (!Array.isArray(blocks)) {
    throw new Error('FoodBlock: blocks must be an array')
  }

  const weights = { ...DEFAULT_WEIGHTS, ...(policy.weights || {}) }
  const now = Date.now()

  const inputs = {
    authority_certs: countAuthorityCerts(actorHash, blocks, policy.required_authorities),
    peer_reviews: computePeerReviews(actorHash, blocks),
    chain_depth: computeChainDepth(actorHash, blocks),
    verified_orders: countVerifiedOrders(actorHash, blocks),
    account_age: computeAccountAge(actorHash, blocks, now)
  }

  const score =
    (inputs.authority_certs * weights.authority_certs) +
    (inputs.peer_reviews.weighted_score * weights.peer_reviews) +
    (inputs.chain_depth * weights.chain_depth) +
    (inputs.verified_orders * weights.verified_orders) +
    (inputs.account_age * weights.account_age)

  const minScore = policy.min_score || 0
  const meetsMinimum = score >= minScore

  return { score, inputs, meets_minimum: meetsMinimum }
}

/**
 * Count valid authority certifications for an actor.
 */
function countAuthorityCerts(actorHash, blocks, requiredAuthorities) {
  let count = 0
  for (const b of blocks) {
    if (b.type !== 'observe.certification') continue
    if (b.refs?.subject !== actorHash) continue
    if (b.state?.valid_until && new Date(b.state.valid_until) < new Date()) continue
    count++
  }
  return count
}

/**
 * Compute peer review score with independence weighting.
 * Returns { count, avg_score, weighted_score }.
 */
function computePeerReviews(actorHash, blocks) {
  const reviews = []
  for (const b of blocks) {
    if (b.type !== 'observe.review') continue
    if (b.refs?.subject !== actorHash) continue
    if (typeof b.state?.rating !== 'number') continue
    reviews.push(b)
  }

  if (reviews.length === 0) {
    return { count: 0, avg_score: 0, weighted_score: 0 }
  }

  let totalWeighted = 0
  let totalWeight = 0

  for (const review of reviews) {
    const reviewerHash = review.refs?.author || review.author_hash
    const density = connectionDensity(reviewerHash, actorHash, blocks)
    const weight = 1 - density
    totalWeighted += (review.state.rating / 5.0) * weight
    totalWeight += weight
  }

  const avgScore = reviews.reduce((sum, r) => sum + r.state.rating, 0) / reviews.length
  const weightedScore = totalWeight > 0 ? totalWeighted / totalWeight * reviews.length : 0

  return { count: reviews.length, avg_score: avgScore, weighted_score: weightedScore }
}

/**
 * Compute effective chain depth: count of distinct author_hash values
 * in the actor's provenance chains (blocks that reference this actor).
 */
function computeChainDepth(actorHash, blocks) {
  const authors = new Set()
  for (const b of blocks) {
    if (!b.refs) continue
    const refsActor = Object.values(b.refs).some(v =>
      v === actorHash || (Array.isArray(v) && v.includes(actorHash))
    )
    if (refsActor && b.author_hash) {
      authors.add(b.author_hash)
    }
  }
  return authors.size
}

/**
 * Count verified orders (transfer.order with adapter_ref).
 */
function countVerifiedOrders(actorHash, blocks) {
  let count = 0
  for (const b of blocks) {
    if (!b.type?.startsWith('transfer.order')) continue
    if (b.refs?.buyer !== actorHash && b.refs?.seller !== actorHash) continue
    if (b.state?.adapter_ref || b.state?.payment_ref) count++
  }
  return count
}

/**
 * Compute account age in days (capped at 365).
 */
function computeAccountAge(actorHash, blocks, now) {
  for (const b of blocks) {
    if (b.hash === actorHash && b.created_at) {
      const created = new Date(b.created_at).getTime()
      const days = (now - created) / (1000 * 60 * 60 * 24)
      return Math.min(days, 365)
    }
  }
  return 0
}

/**
 * Measure connection density between two actors (Section 6.3 sybil resistance).
 * Returns 0..1 where 0 = no shared refs, 1 = fully connected.
 */
function connectionDensity(actorA, actorB, blocks) {
  if (!actorA || !actorB) return 0

  const refsA = new Set()
  const refsB = new Set()

  for (const b of blocks) {
    if (!b.refs) continue
    const vals = Object.values(b.refs).flat()
    if (vals.includes(actorA)) {
      for (const v of vals) {
        if (v !== actorA) refsA.add(v)
      }
    }
    if (vals.includes(actorB)) {
      for (const v of vals) {
        if (v !== actorB) refsB.add(v)
      }
    }
  }

  if (refsA.size === 0 || refsB.size === 0) return 0

  let shared = 0
  for (const ref of refsA) {
    if (refsB.has(ref)) shared++
  }

  const union = new Set([...refsA, ...refsB]).size
  return union > 0 ? shared / union : 0
}

/**
 * Create a trust policy block.
 *
 * @param {string} name - Policy name
 * @param {object} weights - Custom weights
 * @param {object} [opts] - { required_authorities, min_score, author }
 * @returns {object} The trust policy FoodBlock
 */
function createTrustPolicy(name, weights, opts = {}) {
  const { create } = require('./block')

  const state = { name, weights }
  if (opts.required_authorities) state.required_authorities = opts.required_authorities
  if (opts.min_score !== undefined) state.min_score = opts.min_score

  const refs = {}
  if (opts.author) refs.author = opts.author

  return create('observe.trust_policy', state, refs)
}

module.exports = { computeTrust, connectionDensity, createTrustPolicy, DEFAULT_WEIGHTS }
